### ISSUE TITLE
ArgumentError: Invalid byte sequence in UTF-8

### ISSUE THREAD

**Issue by james-ai** [ffi/ffi#792]
Hi there,

I get an ArgumentError when I load my Rails app.

OS: Amazon Linux 2018
Ruby: ruby 2.5.7p206 (2019-10-01 revision 67816) [x86_64-linux]
Rails: 5
ffi gem: 1.12.2

From **/var/log/puma/puma.log**:
```
opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:118:in `rescue in block (2 levels) in ffi_lib': invalid byte sequence in UTF-8 (ArgumentError)
        from /opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:110:in `block (2 levels) in ffi_lib'
        from /opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:109:in `each'
        from /opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:109:in `block in ffi_lib'
        from /opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:99:in `map'
        from /opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb:99:in `ffi_lib'
        from /var/app/current/lib/go_distribution.rb:3:in `<module:GoDistribution>'
        from /var/app/current/lib/go_distribution.rb:1:in `<top (required)>'
```

In `distribution.so` I have:
```ruby
module GoDistribution
  extend FFI::Library
  ffi_lib "#{Rails.root.join('lib', 'ext', 'distribution.so')}"
  attach_function :negbinomdist_native, [:double, :double, :double], :double
  attach_function :poisson_native, [:double, :double, :bool], :double
end
```

The file exists at the location so I'm not sure why ffi can't find it there. Any thoughts?

**Comment by larskanis** [comment]
@james-ai The file `distribution.so` can not be loaded for some reason, but unfortunately the correct error message is hidden behind the UTF-8 error.

Until ffi-1.13.2 ships the fix from #793, you might change the line 118 in `/opt/rubies/ruby-2.5.7/lib/ruby/gems/2.5.0/gems/ffi-1.12.2/lib/ffi/library.rb` to `File.binread` to get the original error. Maybe the CPU architecture of `distribution.so` is wrong or something like, but the correct error message should show you the details.

**Comment by james-ai** [comment]
Hi @larskanis. Thank you for looking at this so quickly. You were correct, and there was another error! In this case it was an ELF header issue. Changed how we build the binaries so hopefully all will be well now.

Thanks,
James.
